### PR TITLE
Documentation for writing custom activations in tiny-dnn.

### PR DESCRIPTION
**Pull request by karandesai-96**
This PR adds a developers' documentation to add a custom activation function to the framework by implementing it as a layer. The documentation clearly explains the methods which a user is supposed to override.

### CODE REVIEW

**Comment by karandesai-96** [comment]
Still a little WIP. I will add in details to write corresponding serialization helpers as well.

File: docs/developer_guides/Writing-Custom-Activations.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+This method contains the main logic of our activation function. It takes
+in two vectors passed by reference and fills the second one by applying 
+activation function to the first one.
+
+For example, let our activation function be simply a scalar multiplication.
+The implementation would look like:
+
+```cpp
+void forward_activation(const vec_t &x, vec_t &y) override {
+  for (serial_size_t j = 0; j < x.size(); j++) {
```
Comment (edgarriba) [inline docs/developer_guides/Writing-Custom-Activations.md]: `size_t j = 0`

File: docs/developer_guides/Writing-Custom-Activations.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    y[j] = alpha * x[j];
+  }
+}
+```
+
+We could have easily accepted ``float_t`` arguments and applied activation 
+function on one element. But this function would have been called for each 
+neuron of the layer. **Calling a virtual function inside a tight for loop 
+hurts performance.** Hence this is how the method is implemented.  
+Practically, each ``vec_t`` here will represent a single flattened image out
```
Comment (edgarriba) [inline docs/developer_guides/Writing-Custom-Activations.md]: flattened Tensor

File: docs/developer_guides/Writing-Custom-Activations.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+For example, the ``backward_activation`` method for our activation function
+would look like:
+
+```cpp
+void backward_activation(const vec_t &x,
+                         const vec_t &y,
+                         vec_t &dx,
+                         const vec_t &dy) override {
+  for (serial_size_t j = 0; j < x.size(); j++) {
```
Comment (edgarriba) [inline docs/developer_guides/Writing-Custom-Activations.md]: `size_t`

**Comment by edgarriba** [comment]
@beru I see, let's fix it in another PR. Merging this.

**Comment by edgarriba** [comment]
should we merge this branch to master?

**Comment by karandesai-96** [comment]
@edgarriba I realized that I made a small mistake. I will have to rename `tanh` to `tan_h` because it is conflicting with built in `tanh`. I am about to submit a PR for the same. Meanwhile other open PRs can be merged into `feat/decouple_activations`.

**Pull request merged**